src/corelibs/U2View/src/ov_msa/AlignSequencesToAlignment/AlignSequencesToAlignmentSupport.cpp

```diff
@@ -0,0 +1,126 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+#include "AlignSequencesToAlignmentSupport.h"
+
+#include <U2Core/AppContext.h>
+#include <U2Core/GObjectSelection.h>
+#include <U2Core/Task.h>
+#include <U2Core/TaskWatchdog.h>
+
+#include <U2Gui/DialogUtils.h>
+#include <U2Gui/LastUsedDirHelper.h>
+#include <U2Gui/ProjectView.h>
+#include <U2Gui/U2FileDialog.h>
+
+#include <U2View/MSAEditor.h>
+#include <U2View/MaEditorFactory.h>
+
+#include "AlignSequencesToAlignmentTask.h"
+
+namespace U2 {
+
+AlignSequencesToAlignmentSupport::AlignSequencesToAlignmentSupport(QObject *parent)
+    : GObjectViewWindowContext(parent, MsaEditorFactory::ID) {
+}
+
+void AlignSequencesToAlignmentSupport::initViewContext(GObjectView *view) {
+    auto msaEditor = qobject_cast<MSAEditor *>(view);
+    SAFE_POINT(msaEditor != nullptr, "View is not MSAEditor!", );
+    CHECK(msaEditor->getMaObject() != nullptr, );
+
+    // TODO: using a single hardcoded (MAFFT) today (this is the original UGENE behavior).
+    //  The final code should build action per each algorithm registered in the AlignmentAlgorithmsRegistry.
+    auto alignAction = new AlignSequencesToAlignmentAction(this, msaEditor, tr("Add sequences with MAFFT..."), 100);
+    alignAction->setIcon(QIcon(":/core/images/add_to_alignment.png"));    //TODO: add a dedicated icon for MAFFT.
+    alignAction->setObjectName("Align with MAFFT");
+    alignAction->setMenuTypes({MsaEditorMenuType::ALIGN_SEQUENCES_TO_ALIGNMENT});
+    alignAction->sl_updateState();
+
+    connect(alignAction, SIGNAL(triggered()), SLOT(sl_alignSequencesToAlignment()));
+    addViewAction(alignAction);
+}
+
+void AlignSequencesToAlignmentSupport::sl_alignSequencesToAlignment() {
+    auto action = qobject_cast<AlignSequencesToAlignmentAction *>(sender());
+    SAFE_POINT(action != nullptr, "Not a AlignSequencesToAlignmentAction!", );
+    auto msaEditor = action->getEditor();
+    auto msaObject = msaEditor->getMaObject();
+    CHECK(!msaObject->isStateLocked(), );
+
+    ProjectView *projectView = AppContext::getProjectView();
+    SAFE_POINT(projectView != nullptr, "Project view is null", );
+
+    const GObjectSelection *selection = projectView->getGObjectSelection();
+    SAFE_POINT(selection != nullptr, "GObjectSelection is null", );
+
+    QList<GObject *> objects = selection->getSelectedObjects();
+    bool selectFromProject = !objects.isEmpty();
+
+    for (const GObject *object : qAsConst(objects)) {
+        const GObjectType &objectType = object->getGObjectType();
+        if (object == msaObject || (objectType != GObjectTypes::MULTIPLE_SEQUENCE_ALIGNMENT && objectType != GObjectTypes::SEQUENCE)) {
+            selectFromProject = false;
+            break;
+        }
+    }
+
+    // TODO: ask user if to select from the project or from files.
+    if (selectFromProject) {
+        SequenceObjectsExtractor extractor;
```

Comment:
I think, it's better to log this branch - it could be not obvious where the aligned sequence was got from

Reply:
Same here: this is the unchanged original code moved out of MSAEditor.cpp into this class.
I think we should not rely on the log at all in UGENE and use very explicit UX. I added a "TODO" comment above.